src/graph.cc

```diff
@@ -33,12 +34,44 @@ bool Node::Stat(DiskInterface* disk_interface, string* err) {
 }
 
 bool DependencyScan::RecomputeDirty(Node* node, string* err) {
-  vector<Node*> stack;
-  return RecomputeDirty(node, &stack, err);
+  std::vector<Node*> stack;
+  std::vector<Node*> validation_nodes;
+  return RecomputeNodeDirty(node, &stack, &validation_nodes, err);
 }
 
-bool DependencyScan::RecomputeDirty(Node* node, vector<Node*>* stack,
+bool DependencyScan::RecomputeDirty(Node* initial_node,
+                                    std::vector<Node*>* validation_nodes,
                                     string* err) {
+  std::vector<Node*> stack;
+  std::vector<Node*> new_validation_nodes;
+
+  std::deque<Node*> nodes;
+  nodes.push_back(initial_node);
+
+  // RecomputeNodeDirty might return new validation nodes that need to be
+  // checked for dirty state, keep a queue of nodes to visit.
+  while (!nodes.empty()) {
+    Node *node = nodes.front();
+    nodes.pop_front();
+
+    stack.clear();
```

Comment:
I would move the declarations here. (Or do you think the performance difference will be noticeable?)

Reply:
In the Android tree we might have thousands of active validation nodes if we use them for Android Lint, Clang tidy and various other checks.  That would mean thousands of times through the loop, and each time through the loop might need to grow the stack a few times.  Reusing the existing stack seems worthwhile.